jaxrs/src/main/java/io/airlift/jaxrs/AsyncResponseHandler.java

```diff
@@ -23,47 +23,56 @@
 import java.lang.ref.WeakReference;
 import java.util.concurrent.Executor;
 import java.util.concurrent.Future;
+import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.function.Supplier;
 
 import static com.google.common.base.Preconditions.checkArgument;
 import static jakarta.ws.rs.core.Response.status;
 import static java.util.Objects.requireNonNull;
 import static java.util.concurrent.TimeUnit.MILLISECONDS;
 
-public class AsyncResponseHandler
+public class AsyncResponseHandler<T>
+        implements FutureCallback<T>
 {
+    private final AtomicBoolean clientDisconnected = new AtomicBoolean();
     private final AsyncResponse asyncResponse;
-    private final WeakReference<Future<?>> futureResponseReference;
+    private final WeakReference<Future<T>> futureResponseReference;
 
-    private AsyncResponseHandler(AsyncResponse asyncResponse, ListenableFuture<?> futureResponse)
+    private AsyncResponseHandler(AsyncResponse asyncResponse, ListenableFuture<T> futureResponse, Executor httpResponseExecutor)
     {
         this.asyncResponse = requireNonNull(asyncResponse, "asyncResponse is null");
         // the jaxrs implementation can hold on to the async timeout for a long time, and
         // the future can reference large expensive objects.  Since we are only interested
         // in canceling this future on a timeout, only hold a weak reference to the future
         this.futureResponseReference = new WeakReference<>(requireNonNull(futureResponse, "futureResponse is null"));
+        Futures.addCallback(futureResponse, this, httpResponseExecutor);
     }
 
-    public static AsyncResponseHandler bindAsyncResponse(AsyncResponse asyncResponse, ListenableFuture<?> futureResponse, Executor httpResponseExecutor)
+    public static <T> AsyncResponseHandler<T> bindAsyncResponse(AsyncResponse asyncResponse, ListenableFuture<T> futureResponse, Executor httpResponseExecutor)
     {
-        Futures.addCallback(futureResponse, toFutureCallback(asyncResponse), httpResponseExecutor);
-        return new AsyncResponseHandler(asyncResponse, futureResponse);
+        return new AsyncResponseHandler<>(asyncResponse, futureResponse, httpResponseExecutor);
     }
 
-    public AsyncResponseHandler withTimeout(Duration timeout)
+    public AsyncResponseHandler<T> withTimeout(Duration timeout)
     {
         return withTimeout(timeout,
                 status(Response.Status.SERVICE_UNAVAILABLE)
                         .entity("Timed out after waiting for " + timeout.convertToMostSuccinctTimeUnit())
                         .build());
     }
 
-    public AsyncResponseHandler withTimeout(Duration timeout, Response timeoutResponse)
+    public void clientDisconnected()
```

Comment:
Move this method below the other `withTimeout()` methods so that the diff is cleaner